libs/@local/harpc/net/src/session/server/connection/mod.rs

```diff
@@ -0,0 +1,342 @@
+        let sink = self.sink;
+        pin!(sink);
+
+        let forward = ReceiverStream::new(self.rx).map(Ok).forward(sink).fuse();
+
+        // redirect the receiver stream to the sink, needs an extra task to drive both
+        select! {
+            result = forward => result,
+            () = cancel.cancelled() => Ok(()),
+        }
+    }
+}
+
+pub(crate) struct ConnectionTask<E> {
+    pub(crate) peer: PeerId,
+    pub(crate) session: SessionId,
+
+    pub(crate) transactions: TransactionCollection,
+    pub(crate) output: mpsc::Sender<Transaction>,
+    pub(crate) events: broadcast::Sender<SessionEvent>,
+
+    pub(crate) config: SessionConfig,
+    pub(crate) encoder: E,
+    pub(crate) _permit: OwnedSemaphorePermit,
+}
+
+impl<E> ConnectionTask<E>
+where
+    E: ErrorEncoder + Send + Sync + 'static,
+{
+    async fn respond_error<T>(&self, id: RequestId, error: T, tx: &mpsc::Sender<Response>)
+    where
+        T: PlainError + Send + Sync,
+    {
+        let TransactionError { code, bytes } = self.encoder.encode_error(error).await;
+
+        let mut writer = ResponseWriter::new(
+            WriterOptions { no_delay: false },
+            ResponseContext {
+                id,
+                kind: ResponseKind::Err(code),
+            },
+            tx,
+        );
+        writer.push(bytes);
+
+        // we cannot stop processing if the writer fails, as that simply means that we can no longer
+        // send and we're winding down.
+        if writer.flush().await.is_err() {
+            tracing::info!("connection prematurely closed, response to transaction has been lost");
+        }
+    }
+
+    async fn handle_request(
+        &self,
+        tx: mpsc::Sender<Response>,
+        tasks: &TaskTracker,
+        request: Request,
+    ) {
+        // check if this is a `Begin` request, in that case we need to create a new transaction,
+        // otherwise, this is already a transaction and we need to forward it, or log out if it is a
+        // rogue request
+        let request_id = request.header.request_id;
+
+        // these transactions then need to be propagated to the main session layer via an mpsc
+        // channel, which drops a transaction if there's too many.
+        match &request.body {
+            RequestBody::Begin(begin) => {
+                #[expect(
+                    clippy::significant_drop_in_scrutinee,
+                    reason = "This simply returns a drop guard, that is carried through the \
+                              transaction lifetime."
+                )]
+                let (permit, request_tx, request_rx) =
+                    match self.transactions.acquire(request_id).await {
+                        Ok((permit, tx, rx)) => (permit, tx, rx),
+                        Err(error) => {
+                            tracing::info!("transaction limit reached, dropping transaction");
+
+                            return self.respond_error(request_id, error, &tx).await;
```

Comment:
`respond_error` does return `()`, using a `return` here is misleading.
```suggestion
                            self.respond_error(request_id, error, &tx).await;
                            return;
```

Reply:
Yes, this is an überbleibsel from a previous version, where `self.respond_error` did actually return something. Thank you!